File: Java/singly linkedlist.java
public class LinkedListExample {

    // Node class
    static class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
        }
    }

    // Insert node at end
    public static Node insert(Node head, int val) {
        Node newNode = new Node(val);
        if (head == null) return newNode;

        Node temp = head;
        while (temp.next != null)
            temp = temp.next;
        temp.next = newNode;
        return head;
    }

    // Reverse linked list
    public static Node reverse(Node head) {
        Node prev = null, curr = head;
        while (curr != null) {
            Node next = curr.next;  // Store next
            curr.next = prev;       // Reverse pointer
            prev = curr;
            curr = next;
        }
        return prev;
    }

    // Print list
    public static void printList(Node head) {
        while (head != null) {
            System.out.print(head.data + " ");
            head = head.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Node head = null;

        head = insert(head, 1);
        head = insert(head, 2);
        head = insert(head, 3);

        System.out.print("Original list: ");
        printList(head);  // 1 2 3

        head = reverse(head);
        System.out.print("Reversed list: ");
        printList(head);  // 3 2 1
    }
}

File: C++/singly linkedlist.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;

    Node(int val) {
        data = val;
        next = nullptr;
    }
};

void insertAtEnd(Node*& head, int val) {
    Node* newNode = new Node(val);

    if (!head) {
        head = newNode;
        return;
    }

    Node* curr = head;
    while (curr->next) curr = curr->next;
    curr->next = newNode;
}

void reverse(Node*& head) {
    Node* prev = nullptr;
    Node* curr = head;
    Node* next;

    while (curr) {
        next = curr->next;     // Store next
        curr->next = prev;     // Reverse pointer
        prev = curr;           // Move prev forward
        curr = next;           // Move curr forward
    }
    head = prev;
}

void printList(Node* head) {
    while (head) {
        cout << head->data << " ";
        head = head->next;
    }
}

// ✅ Use-case
int main() {
    Node* head = nullptr;
    insertAtEnd(head, 10);
    insertAtEnd(head, 20);
    insertAtEnd(head, 30);

    cout << "Original List: ";
    printList(head);  // 10 20 30

    reverse(head);
    cout << "\nReversed List: ";
    printList(head);  // 30 20 10
}
